### PR TITLE
Make msgpack-java an optional dependency

### PR DESCRIPTION
**Pull request by bioball**
Make msgpack-java a compileTime dependency in pkl-core, and include it:

* In the fat jars (pkl-config-java-all, pkl-conifg-java)
* In the CLIs

The is the alternative to https://github.com/apple/pkl/pull/763

### CODE REVIEW

**Comment by bioball** [comment]
Alternative PR: https://github.com/apple/pkl/pull/763

File: pkl-cli/gradle.lockfile [hunk 1]
```diff
@@ -74,10 +74,14 @@ org.jetbrains.kotlin:kotlin-scripting-common:1.7.10=kotlinCompilerPluginClasspat
...
+org.jetbrains.kotlin:kotlin-stdlib-common:1.8.22=compileClasspath,runtimeClasspath,testCompileClasspath,testRuntimeClasspath
+org.jetbrains.kotlin:kotlin-stdlib-jdk7:1.7.10=apiDependenciesMetadata,implementationDependenciesMetadata,testImplementationDependenciesMetadata
+org.jetbrains.kotlin:kotlin-stdlib-jdk7:1.8.22=compileClasspath,runtimeClasspath,testCompileClasspath,testRuntimeClasspath
+org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.7.10=apiDependenciesMetadata,implementationDependenciesMetadata,testImplementationDependenciesMetadata
+org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.8.22=compileClasspath,runtimeClasspath,testCompileClasspath,testRuntimeClasspath
+org.jetbrains.kotlin:kotlin-stdlib:1.7.10=apiDependenciesMetadata,implementationDependenciesMetadata,kotlinCompilerClasspath,kotlinCompilerPluginClasspathMain,kotlinCompilerPluginClasspathTest,kotlinKlibCommonizerClasspath,testImplementationDependenciesMetadata
+org.jetbrains.kotlin:kotlin-stdlib:1.8.22=compileClasspath,runtimeClasspath,testCompileClasspath,testRuntimeClasspath
```
Comment (holzensp) [inline pkl-cli/gradle.lockfile:84]: Let's avoid multiple Kotlin stdlib version dependencies.

File: pkl-core/src/main/java/org/pkl/core/module/ModuleKeyFactories.java [hunk 2]
```diff
@@ -79,7 +80,11 @@ public static ModuleKeyFactory classPath(ClassLoader classLoader) {
   }
 
   /**
-   * Returns a factory for external reader module keys.
+   * Returns a factory for external process module keys.
+   *
+   * <p>Using external process module keys requires a separate runtime dependency on <a
+   * href="https://github.com/msgpack/msgpack-java">msgpack-java</a> version 0.9.x, unless using
```
Comment (holzensp) [inline pkl-core/src/main/java/org/pkl/core/module/ModuleKeyFactories.java:86]: Not a big fan of literal version numbers in doc-comments for risk of going out of sync.

File: pkl-core/src/main/java/org/pkl/core/module/ModuleKeyFactories.java [hunk 3]
```diff
@@ -89,7 +94,11 @@ public static ModuleKeyFactory externalProcess(String scheme, ExternalReaderProc
...
```
Comment (holzensp) [inline pkl-core/src/main/java/org/pkl/core/module/ModuleKeyFactories.java:100]: ^^ literal version number in doc comment

File: pkl-core/src/main/java/org/pkl/core/resource/ResourceReaders.java [hunk 4]
```diff
@@ -143,6 +143,10 @@ public static List<ResourceReader> fromServiceProviders() {
   /**
    * Returns a reader for external reader resources.
    *
+   * <p>Using external process resource readers requires a separate runtime dependency on <a
+   * href="https://github.com/msgpack/msgpack-java">msgpack-java</a> version 0.9.x, unless using
```
Comment (holzensp) [inline pkl-core/src/main/java/org/pkl/core/resource/ResourceReaders.java:147]: ^^ literal version number in doc comment

File: pkl-core/src/main/java/org/pkl/core/resource/ResourceReaders.java [hunk 5]
```diff
@@ -153,6 +157,10 @@ public static ResourceReader externalProcess(String scheme, ExternalReaderProces
...
```
Comment (holzensp) [inline pkl-core/src/main/java/org/pkl/core/resource/ResourceReaders.java:161]: ^^ literal version number in doc comment

**Review by holzensp — APPROVED** [review]
Do please remove literal version numbers in doc comments. More importantly; please resolve Kotlin stdlib versions. Approving to unblock; this is my preferred solution, over #763.

(CI failure is Spotless)

**Comment by bioball** [comment]
Closing this; adding this introduces a possible runtime exception that doesn't seem to be worth it.

**Pull request closed without merging**